### PR TITLE
pool - initial commit

### PR DESCRIPTION
**Pull request by vetinari**
the pooling is adapted from https://github.com/fatih/pool

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -7,6 +7,7 @@ type Client interface {
 	Start()
 	StartTLS(config *tls.Config) error
 	Close()
+	Alive() bool
```
Comment (liggitt) [inline client.go]: Not sure about this… do other types of net connections have equivalent methods?

File: conn.go [hunk 2]
```diff
@@ -106,6 +106,20 @@ func NewConn(conn net.Conn, isTLS bool) *Conn {
 	}
 }
 
+func (l *Conn) Alive() bool {
```
Comment (liggitt) [inline conn.go]: Is this safe to do on a connection that might be running another request already?

Comment (liggitt) [inline conn.go]: Also, I'm not positive this will always succeed for connections that are still alive.

Comment (liggitt) [inline conn.go]: Might be better to make the liveness check pluggable and part of the pool

File: pool_conn.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+// MarkUnusable() marks the connection not usable any more, to let the pool close it
+// instead of returning it to pool.
+func (p *PoolConn) MarkUnusable() {
+	p.unusable = true
+}
+
+func (p *PoolConn) autoClose(err error) {
+	for _, code := range p.closeAt {
```
Comment (liggitt) [inline pool_conn.go]: Not seeing where this is called from. Don't all the delegated calls have to pass any resulting errors through here?

File: pool_conn.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+// Close() puts the given connects back to the pool instead of closing it.
+func (p *PoolConn) Close() {
+	if p.unusable {
+		log.Printf("Closing unusable connection %v", p.Conn)
+		if p.Conn != nil {
+			p.Conn.Close()
+		}
+		return
+	}
+	p.c.put(p.Conn)
```
Comment (liggitt) [inline pool_conn.go]: Need to be super careful about retained state here, like leaving a bound connection to be reused

Comment (liggitt) [inline pool_conn.go]: the things I can think of off the top of my head are whether TLS has been started and whether the connection was bound to a dn/password

File: pool_conn.go (hunk 3 above)
Comment (vetinari) [inline pool_conn.go]: ah, that's missing from my other code

**Comment by liggitt** [comment]
I like the idea of a connection pool, but until we do a sweep of the existing request logic to see how an errant response to an earlier request would be handled, I wouldn't want to claim support for reused connections. The TLS and bind state reuse would also need to be thought through to prevent accidental credential reuse

**Comment by vetinari** [comment]
Further development will happen in https://github.com/vetinari/go-ldappool

**Pull request closed without merging**